### ISSUE TITLE
__attribute__((naked)) in ISRs on Cortex-M platforms

### ISSUE THREAD

**Issue by jnohlgard** [RIOT-OS/RIOT#1654]
**TL;DR:** Can `__attribute__((naked))` and `ISR_ENTER()`, `ISR_EXIT()` be safely used when not writing ISRs in assembly language?

Should existing ISRs be converted to normal functions _if they are not written purely in inline assembler_?

Long post below:
# Background

I ran into a problem where I got a hardfault in the idle_thread after servicing an interrupt on an UART. The `ISR_ENTER()` macro used as a function prologue for ISRs in the ARM Cortex ports only pushes the lr register to the stack. The hardware pushes r0-r3, r12, lr to the stack automatically when switching to an ISR, but r4-r11 are expected to be user handled. 
# Analysis

When I removed `__attribute__((naked))` and `ISR_ENTER()`, `ISR_EXIT()` from the ISR (i.e. made it a normal function) the hardfaults went away.

After analyzing the problem I discovered that my ISR was using the r4 CPU register without pushing it to the stack first. GCC generated assembly that would make use of the r4 register but since I was using `__attribute__((naked))`, no function prologue is being generated by GCC.

The automatically generated function prologue pushes r4 to the stack and pops it before returning.
# Question

Should the Cortex M ports code really use `__attribute__((naked))` and `ISR_ENTER()`, `ISR_EXIT()` if the ISR is written in C (or any other language dependent on the compiler to generate assembly language) in order to safeguard against future changes to the compiler code generator?
# Example code

I have attached two pieces of disassembled ISR code below, pay special attention to the instructions at 0 and 2.

Side note: I do not understand the instructions at 3e in the second example, why is r4 and lr loaded from the stack before calling `thread_arch_yield` in the normal function, but not in the naked function?

First one, with `__attribute__((naked))` and `ISR_ENTER()`, `ISR_EXIT()`:

```
00000000 <isr_uart1_status>:
   0:   b500        push    {lr}
   2:   4c11        ldr r4, [pc, #68]   ; (48 <isr_uart1_status+0x48>)
   4:   7923        ldrb    r3, [r4, #4]
   6:   069a        lsls    r2, r3, #26
   8:   d50a        bpl.n   20 <isr_uart1_status+0x20>
   a:   4b10        ldr r3, [pc, #64]   ; (4c <isr_uart1_status+0x4c>)
   c:   79e0        ldrb    r0, [r4, #7]
   e:   f9b3 3000   ldrsh.w r3, [r3]
  12:   b2c0        uxtb    r0, r0
  14:   b183        cbz r3, 38 <isr_uart1_status+0x38>
  16:   f7ff fffe   bl  0 <uart0_handle_incoming>
  1a:   f7ff fffe   bl  0 <uart0_notify_thread>
  1e:   e00b        b.n 38 <isr_uart1_status+0x38>
  20:   7923        ldrb    r3, [r4, #4]
  22:   061b        lsls    r3, r3, #24
  24:   d508        bpl.n   38 <isr_uart1_status+0x38>
  26:   4b0a        ldr r3, [pc, #40]   ; (50 <isr_uart1_status+0x50>)
  28:   685a        ldr r2, [r3, #4]
  2a:   6898        ldr r0, [r3, #8]
  2c:   4790        blx r2
  2e:   b918        cbnz    r0, 38 <isr_uart1_status+0x38>
  30:   78e3        ldrb    r3, [r4, #3]
  32:   f003 037f   and.w   r3, r3, #127    ; 0x7f
  36:   70e3        strb    r3, [r4, #3]
  38:   4b06        ldr r3, [pc, #24]   ; (54 <isr_uart1_status+0x54>)
  3a:   681b        ldr r3, [r3, #0]
  3c:   b10b        cbz r3, 42 <isr_uart1_status+0x42>
  3e:   f7ff fffe   bl  0 <thread_arch_yield>
  42:   bc01        pop {r0}
  44:   4700        bx  r0
  46:   bf00        nop
  48:   4006b000    .word   0x4006b000
```

Second one, normal function, no `ISR_ENTER()`, `ISR_EXIT()`:

```
00000000 <isr_uart1_status>:
   0:   b510        push    {r4, lr}
   2:   4c11        ldr r4, [pc, #68]   ; (48 <isr_uart1_status+0x48>)
   4:   7923        ldrb    r3, [r4, #4]
   6:   069a        lsls    r2, r3, #26
   8:   d50a        bpl.n   20 <isr_uart1_status+0x20>
   a:   4b10        ldr r3, [pc, #64]   ; (4c <isr_uart1_status+0x4c>)
   c:   79e0        ldrb    r0, [r4, #7]
   e:   f9b3 3000   ldrsh.w r3, [r3]
  12:   b2c0        uxtb    r0, r0
  14:   b183        cbz r3, 38 <isr_uart1_status+0x38>
  16:   f7ff fffe   bl  0 <uart0_handle_incoming>
  1a:   f7ff fffe   bl  0 <uart0_notify_thread>
  1e:   e00b        b.n 38 <isr_uart1_status+0x38>
  20:   7923        ldrb    r3, [r4, #4]
  22:   061b        lsls    r3, r3, #24
  24:   d508        bpl.n   38 <isr_uart1_status+0x38>
  26:   4b0a        ldr r3, [pc, #40]   ; (50 <isr_uart1_status+0x50>)
  28:   685a        ldr r2, [r3, #4]
  2a:   6898        ldr r0, [r3, #8]
  2c:   4790        blx r2
  2e:   b918        cbnz    r0, 38 <isr_uart1_status+0x38>
  30:   78e3        ldrb    r3, [r4, #3]
  32:   f003 037f   and.w   r3, r3, #127    ; 0x7f
  36:   70e3        strb    r3, [r4, #3]
  38:   4b06        ldr r3, [pc, #24]   ; (54 <isr_uart1_status+0x54>)
  3a:   681b        ldr r3, [r3, #0]
  3c:   b11b        cbz r3, 46 <isr_uart1_status+0x46>
  3e:   e8bd 4010   ldmia.w sp!, {r4, lr}
  42:   f7ff bffe   b.w 0 <thread_arch_yield>
  46:   bd10        pop {r4, pc}
  48:   4006b000    .word   0x4006b000
```

GCC version: `arm-none-eabi-gcc (Gentoo 4.8.3 p1.1, pie-0.5.9) 4.8.3`
